python/ray/data/dataset_pipeline.py

```diff
@@ -1259,9 +1255,15 @@ def add_stage(ds, stage):
 
     def _peek(self) -> Dataset[T]:
         if self._first_dataset is None:
-            self._optimize_stages()
-            self._dataset_iter = PipelineExecutor(self)
-            self._first_dataset = next(self._dataset_iter)
+            first_dataset_gen = next(iter(self._base_iterable))
+            peek_pipe = DatasetPipeline(
+                base_iterable=[first_dataset_gen],
+                stages=self._stages.copy(),
+                length=1,
+                progress_bars=False,
+            )
```

Comment:
the primary change is here: instead of executed against self, we create a new pipeline of length one that we can peek against.

As before, once we execute the peek once, we cache the peeked dataset so we don't have to execute again for subsequent peeks.

Reply:
Put this as a comment / impl note into the code?

Reply:
added